File: backend/src/services/whatsapp.rs
use crate::error::{AppError, Result};
use serde::Deserialize;

const SIDECAR_URL: &str = "http://127.0.0.1:3099";

#[derive(Deserialize)]
pub struct SidecarStatus {
    pub connected: bool,
    pub qr: Option<String>,
}

/// Send a WhatsApp message via the local Baileys sidecar.
/// `phone` must be in international digits-only format, e.g. "919876543210".
pub async fn send_message(phone: &str, body: &str) -> Result<()> {
    let normalised = normalise_phone(phone);

    let client = reqwest::Client::new();
    let res = client
        .post(format!("{SIDECAR_URL}/api/send"))
        .json(&serde_json::json!({ "phone": normalised, "message": body }))
        .send()
        .await
        .map_err(|_| AppError::WhatsApp(
            "Could not reach the local WhatsApp service. Make sure the sidecar is running.".into()
        ))?;

    if res.status().as_u16() == 503 {
        return Err(AppError::WhatsApp(
            "WhatsApp is not connected. Please scan the QR code in Settings → WhatsApp.".into()
        ));
    }

    if !res.status().is_success() {
        let err: serde_json::Value = res.json().await.unwrap_or_default();
        let msg = err["error"].as_str().unwrap_or("Unknown sidecar error").to_string();
        return Err(AppError::WhatsApp(msg));
    }

    Ok(())
}

/// Proxy the sidecar's status (connected + QR data URL) to the frontend.
pub async fn get_status() -> Result<SidecarStatus> {
    let client = reqwest::Client::new();
    let res = client
        .get(format!("{SIDECAR_URL}/api/status"))
        .send()
        .await
        .map_err(|_| AppError::WhatsApp(
            "Could not reach the local WhatsApp service. Make sure the sidecar is running.".into()
        ))?;

    let status: SidecarStatus = res.json().await
        .map_err(|e| AppError::WhatsApp(e.to_string()))?;

    Ok(status)
}

/// Disconnect WhatsApp and wipe saved credentials.
pub async fn logout() -> Result<()> {
    let client = reqwest::Client::new();
    client
        .post(format!("{SIDECAR_URL}/api/logout"))
        .send()
        .await
        .map_err(|_| AppError::WhatsApp("Could not reach the local WhatsApp service.".into()))?;
    Ok(())
}

/// Normalise an Indian phone number to digits-only international format.
/// 10-digit → "91" + digits
/// Leading "0" + 10 digits → "91" + digits
/// Already has "91" prefix → unchanged
fn normalise_phone(phone: &str) -> String {
    let digits: String = phone.chars().filter(|c| c.is_ascii_digit()).collect();
    if digits.starts_with("91") && digits.len() == 12 {
        digits
    } else if digits.starts_with('0') && digits.len() == 11 {
        format!("91{}", &digits[1..])
    } else if digits.len() == 10 {
        format!("91{digits}")
    } else {
        digits
    }
}

File: next-app/src/lib/services/whatsapp.ts
import { env } from "@/lib/env";
import { AppError } from "@/lib/errors";

export interface SidecarStatus {
  connected: boolean;
  qr: string | null;
}

const UNREACHABLE = "Could not reach the local WhatsApp service. Make sure the sidecar is running.";
const UNAUTHORIZED = "WhatsApp service rejected the request. Check that SIDECAR_TOKEN matches the sidecar.";

/** Headers for every sidecar call, including the shared-secret token. */
function sidecarHeaders(extra: Record<string, string> = {}): Record<string, string> {
  return env.SIDECAR_TOKEN ? { ...extra, "x-sidecar-token": env.SIDECAR_TOKEN } : extra;
}

/**
 * Normalise an Indian phone number to digits-only international format.
 * 10-digit → "91" + digits
 * Leading "0" + 10 digits → "91" + digits
 * Already has "91" prefix → unchanged
 */
function normalisePhone(phone: string): string {
  const digits = phone.replace(/\D/g, "");
  if (digits.startsWith("91") && digits.length === 12) return digits;
  if (digits.startsWith("0") && digits.length === 11) return `91${digits.slice(1)}`;
  if (digits.length === 10) return `91${digits}`;
  return digits;
}

/** Send a WhatsApp message via the local Baileys sidecar. */
export async function sendMessage(phone: string, body: string): Promise<void> {
  const normalised = normalisePhone(phone);

  let res: Response;
  try {
    res = await fetch(`${env.SIDECAR_URL}/api/send`, {
      method: "POST",
      headers: sidecarHeaders({ "Content-Type": "application/json" }),
      body: JSON.stringify({ phone: normalised, message: body }),
    });
  } catch {
    throw AppError.whatsapp(UNREACHABLE);
  }

  if (res.status === 401) throw AppError.whatsapp(UNAUTHORIZED);

  if (res.status === 503) {
    throw AppError.whatsapp("WhatsApp is not connected. Please scan the QR code in Settings → WhatsApp.");
  }

  if (!res.ok) {
    const err = await res.json().catch(() => ({}));
    throw AppError.whatsapp(err.error ?? "Unknown sidecar error");
  }
}

/** Proxy the sidecar's status (connected + QR data URL) to the frontend. */
export async function getStatus(): Promise<SidecarStatus> {
  let res: Response;
  try {
    res = await fetch(`${env.SIDECAR_URL}/api/status`, { headers: sidecarHeaders() });
  } catch {
    throw AppError.whatsapp(UNREACHABLE);
  }
  if (res.status === 401) throw AppError.whatsapp(UNAUTHORIZED);
  return res.json();
}

/** Disconnect WhatsApp and wipe saved credentials. */
export async function logout(): Promise<void> {
  let res: Response;
  try {
    res = await fetch(`${env.SIDECAR_URL}/api/logout`, { method: "POST", headers: sidecarHeaders() });
  } catch {
    throw AppError.whatsapp("Could not reach the local WhatsApp service.");
  }
  if (res.status === 401) throw AppError.whatsapp(UNAUTHORIZED);
}
